File: packages/sardis-mcp-server/src/tools/mpp.ts
/**
 * Sardis MCP Server - MPP (Machine Payments Protocol) Tools
 *
 * Tools for managing MPP payment sessions on Tempo.
 */

import { z } from 'zod';
import type { ToolDefinition, ToolHandler, ToolResult } from './types.js';
import { apiRequest } from '../api.js';
import { getConfig } from '../config.js';

const CreateSessionSchema = z.object({
  spending_limit: z.number().positive().describe('Maximum spending for this session in USD'),
  method: z.enum(['tempo', 'stripe_spt']).optional().default('tempo'),
  chain: z.string().optional().default('tempo'),
  currency: z.string().optional().default('USDC'),
  mandate_id: z.string().optional().describe('Spending mandate to bind to'),
  wallet_id: z.string().optional().describe('Wallet ID to use'),
  agent_id: z.string().optional().describe('Agent ID for this session'),
  expires_in_seconds: z.number().int().positive().optional().default(3600),
});

const ExecutePaymentSchema = z.object({
  session_id: z.string().describe('MPP session ID'),
  amount: z.number().positive().describe('Payment amount'),
  merchant: z.string().describe('Merchant identifier or URL'),
  merchant_url: z.string().optional().describe('Merchant URL for MPP discovery'),
  memo: z.string().optional().describe('Payment memo'),
});

const SessionIdSchema = z.object({
  session_id: z.string().describe('MPP session ID'),
});

const EvaluatePolicySchema = z.object({
  amount: z.number().positive().describe('Payment amount to evaluate'),
  merchant: z.string().describe('Merchant identifier'),
  payment_type: z.string().optional().default('mpp_tempo'),
  currency: z.string().optional().default('USDC'),
  network: z.string().optional().default('tempo'),
});

function serialize(result: unknown): ToolResult {
  return { content: [{ type: 'text', text: JSON.stringify(result, null, 2) }] };
}

function errorResult(message: string): ToolResult {
  return { content: [{ type: 'text', text: message }], isError: true };
}

const IssueCardSchema = z.object({
  amount: z.number().min(5).max(1000).describe('Card amount in USD ($5-$1,000)'),
  currency: z.string().optional().default('USD'),
  session_id: z.string().optional().describe('MPP session to charge against'),
});

export const mppToolDefinitions: ToolDefinition[] = [
  {
    name: 'sardis_mpp_create_session',
    description:
      'Create an MPP payment session with a spending limit for agent payments on Tempo. Returns session ID for subsequent payments.',
    inputSchema: {
      type: 'object',
      properties: {
        spending_limit: {
          type: 'number',
          description: 'Maximum spending for this session in USD',
        },
        method: {
          type: 'string',
          enum: ['tempo', 'stripe_spt'],
          description: 'Payment method (default: tempo)',
        },
        chain: {
          type: 'string',
          description: 'Target chain (default: tempo)',
        },
        currency: {
          type: 'string',
          description: 'Payment currency (default: USDC)',
        },
        mandate_id: {
          type: 'string',
          description: 'Spending mandate to bind this session to',
        },
        wallet_id: {
          type: 'string',
          description: 'Wallet ID to use for payments',
        },
        agent_id: {
          type: 'string',
          description: 'Agent ID for this session',
        },
        expires_in_seconds: {
          type: 'number',
          description: 'Session TTL in seconds (default: 3600)',
        },
      },
      required: ['spending_limit'],
    },
  },
  {
    name: 'sardis_mpp_execute',
    description:
      'Execute an MPP payment within an active session. Validates against session spending limit before processing.',
    inputSchema: {
      type: 'object',
      properties: {
        session_id: {
          type: 'string',
          description: 'MPP session ID from sardis_mpp_create_session',
        },
        amount: {
          type: 'number',
          description: 'Payment amount in session currency',
        },
        merchant: {
          type: 'string',
          description: 'Merchant identifier or domain (e.g., "openai.com")',
        },
        merchant_url: {
          type: 'string',
          description: 'Full merchant URL for MPP service discovery',
        },
        memo: {
          type: 'string',
          description: 'Payment memo for audit trail',
        },
      },
      required: ['session_id', 'amount', 'merchant'],
    },
  },
  {
    name: 'sardis_mpp_close_session',
    description: 'Close an MPP session and settle remaining balance. Returns final session summary.',
    inputSchema: {
      type: 'object',
      properties: {
        session_id: {
          type: 'string',
          description: 'MPP session ID to close',
        },
      },
      required: ['session_id'],
    },
  },
  {
    name: 'sardis_mpp_get_session',
    description: 'Get MPP session status including remaining budget and payment count.',
    inputSchema: {
      type: 'object',
      properties: {
        session_id: {
          type: 'string',
          description: 'MPP session ID to query',
        },
      },
      required: ['session_id'],
    },
  },
  {
    name: 'sardis_mpp_issue_card',
    description:
      'Issue a virtual prepaid Visa card via Laso Finance MPP service. Cards are single-use, $5-$1,000, funded with USDC. Can be charged against an MPP session budget.',
    inputSchema: {
      type: 'object',
      properties: {
        amount: {
          type: 'number',
          description: 'Card amount in USD ($5-$1,000). Must match checkout total exactly.',
        },
        currency: {
          type: 'string',
          description: 'Card currency (default: USD)',
        },
        session_id: {
          type: 'string',
          description: 'MPP session ID to charge the card against (optional)',
        },
      },
      required: ['amount'],
    },
  },
  {
    name: 'sardis_mpp_evaluate_policy',
    description: 'Evaluate Sardis spending policy for an MPP payment without executing it (dry-run).',
    inputSchema: {
      type: 'object',
      properties: {
        amount: {
          type: 'number',
          description: 'Payment amount to evaluate',
        },
        merchant: {
          type: 'string',
          description: 'Merchant identifier',
        },
        payment_type: {
          type: 'string',
          description: 'Payment type (default: mpp_tempo)',
        },
        currency: {
          type: 'string',
          description: 'Currency (default: USDC)',
        },
        network: {
          type: 'string',
          description: 'Network (default: tempo)',
        },
      },
      required: ['amount', 'merchant'],
    },
  },
];

export const mppToolHandlers: Record<string, ToolHandler> = {
  sardis_mpp_create_session: async (args: unknown): Promise<ToolResult> => {
    const parsed = CreateSessionSchema.safeParse(args);
    if (!parsed.success) return errorResult(`Invalid request: ${parsed.error.message}`);

    const config = getConfig();

    if (!config.apiKey || config.mode === 'simulated') {
      const sessionId = `mpp_sess_sim_${Date.now().toString(36)}`;
      return serialize({
        session_id: sessionId,
        mandate_id: parsed.data.mandate_id || null,
        wallet_id: parsed.data.wallet_id || config.walletId || null,
        agent_id: parsed.data.agent_id || config.agentId || null,
        method: parsed.data.method,
        chain: parsed.data.chain,
        currency: parsed.data.currency,
        spending_limit: parsed.data.spending_limit.toFixed(2),
        remaining: parsed.data.spending_limit.toFixed(2),
        total_spent: '0.00',
        payment_count: 0,
        status: 'active',
        created_at: new Date().toISOString(),
        expires_at: new Date(Date.now() + (parsed.data.expires_in_seconds ?? 3600) * 1000).toISOString(),
      });
    }

    try {
      const result = await apiRequest('POST', '/api/v2/mpp/sessions', {
        spending_limit: parsed.data.spending_limit,
        method: parsed.data.method,
        chain: parsed.data.chain,
        currency: parsed.data.currency,
        mandate_id: parsed.data.mandate_id,
        wallet_id: parsed.data.wallet_id || config.walletId,
        agent_id: parsed.data.agent_id || config.agentId,
        expires_in_seconds: parsed.data.expires_in_seconds,
      });
      return serialize(result);
    } catch (error) {
      return errorResult(`Failed to create MPP session: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  },

  sardis_mpp_execute: async (args: unknown): Promise<ToolResult> => {
    const parsed = ExecutePaymentSchema.safeParse(args);
    if (!parsed.success) return errorResult(`Invalid request: ${parsed.error.message}`);

    const config = getConfig();

    if (!config.apiKey || config.mode === 'simulated') {
      const paymentId = `mpp_pay_sim_${Date.now().toString(36)}`;
      const remaining = Math.max(0, 100 - parsed.data.amount); // simulated
      return serialize({
        payment_id: paymentId,
        session_id: parsed.data.session_id,
        amount: parsed.data.amount.toFixed(2),
        merchant: parsed.data.merchant,
        status: 'completed',
        tx_hash: '0x' + Math.random().toString(16).substring(2).padEnd(64, '0'),
        chain: 'tempo',
        remaining: remaining.toFixed(2),
      });
    }

    try {
      const result = await apiRequest('POST', `/api/v2/mpp/sessions/${parsed.data.session_id}/execute`, {
        amount: parsed.data.amount,
        merchant: parsed.data.merchant,
        merchant_url: parsed.data.merchant_url,
        memo: parsed.data.memo,
      });
      return serialize(result);
    } catch (error) {
      const msg = error instanceof Error ? error.message : 'Unknown error';
      if (msg.includes('exceeds') || msg.includes('budget') || msg.includes('limit')) {
        return serialize({
          success: false,
          status: 'BLOCKED',
          error: 'SESSION_BUDGET_EXCEEDED',
          message: msg,
          prevention: 'Financial Hallucination PREVENTED',
        });
      }
      return errorResult(`Failed to execute MPP payment: ${msg}`);
    }
  },

  sardis_mpp_close_session: async (args: unknown): Promise<ToolResult> => {
    const parsed = SessionIdSchema.safeParse(args);
    if (!parsed.success) return errorResult(`Invalid request: ${parsed.error.message}`);

    const config = getConfig();

    if (!config.apiKey || config.mode === 'simulated') {
      return serialize({
        session_id: parsed.data.session_id,
        status: 'closed',
        total_spent: '50.00',
        remaining: '50.00',
        payment_count: 2,
        closed_at: new Date().toISOString(),
      });
    }

    try {
      const result = await apiRequest('POST', `/api/v2/mpp/sessions/${parsed.data.session_id}/close`);
      return serialize(result);
    } catch (error) {
      return errorResult(`Failed to close MPP session: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  },

  sardis_mpp_get_session: async (args: unknown): Promise<ToolResult> => {
    const parsed = SessionIdSchema.safeParse(args);
    if (!parsed.success) return errorResult(`Invalid request: ${parsed.error.message}`);

    const config = getConfig();

    if (!config.apiKey || config.mode === 'simulated') {
      return serialize({
        session_id: parsed.data.session_id,
        status: 'active',
        method: 'tempo',
        chain: 'tempo',
        currency: 'USDC',
        spending_limit: '100.00',
        remaining: '75.00',
        total_spent: '25.00',
        payment_count: 1,
        created_at: new Date(Date.now() - 600000).toISOString(),
        expires_at: new Date(Date.now() + 3000000).toISOString(),
      });
    }

    try {
      const result = await apiRequest('GET', `/api/v2/mpp/sessions/${parsed.data.session_id}`);
      return serialize(result);
    } catch (error) {
      return errorResult(`Failed to get MPP session: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  },

  sardis_mpp_issue_card: async (args: unknown): Promise<ToolResult> => {
    const parsed = IssueCardSchema.safeParse(args);
    if (!parsed.success) return errorResult(`Invalid request: ${parsed.error.message}`);

    const config = getConfig();

    if (!config.apiKey || config.mode === 'simulated') {
      const cardId = `card_sim_${Date.now().toString(36)}`;
      return serialize({
        card_id: cardId,
        card_number: '4111XXXXXXXX' + Math.floor(1000 + Math.random() * 9000),
        cvv: '***',
        expiry: '12/27',
        amount: parsed.data.amount.toFixed(2),
        currency: parsed.data.currency,
        status: 'issued',
        card_type: 'single_use',
        provider: 'laso_finance',
        message: `Virtual Visa card issued for $${parsed.data.amount.toFixed(2)} via Laso Finance MPP`,
      });
    }

    try {
      const result = await apiRequest('POST', '/api/v2/mpp/cards/issue', {
        amount: parsed.data.amount,
        currency: parsed.data.currency,
        session_id: parsed.data.session_id,
      });
      return serialize(result);
    } catch (error) {
      return errorResult(`Failed to issue card: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  },

  sardis_mpp_evaluate_policy: async (args: unknown): Promise<ToolResult> => {
    const parsed = EvaluatePolicySchema.safeParse(args);
    if (!parsed.success) return errorResult(`Invalid request: ${parsed.error.message}`);

    const config = getConfig();

    if (!config.apiKey || config.mode === 'simulated') {
      return serialize({
        allowed: true,
        reason: 'ALLOWED by default policy (simulated)',
        checks_passed: 12,
        checks_total: 12,
      });
    }

    try {
      const result = await apiRequest('POST', '/api/v2/mpp/evaluate', {
        amount: parsed.data.amount,
        merchant: parsed.data.merchant,
        payment_type: parsed.data.payment_type,
        currency: parsed.data.currency,
        network: parsed.data.network,
      });
      return serialize(result);
    } catch (error) {
      return errorResult(`Failed to evaluate policy: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  },
};

File: packages/reference-api/server/routes/protocol/mpp.py
"""MPP (Machine Payments Protocol) API endpoints.

Provides session-based MPP payment management:
- Create MPP sessions with spending mandates
- Execute payments within sessions
- Close sessions and settle remaining
- Policy evaluation and dry-run simulation
- Virtual card issuance via Laso Finance

Persistence: Uses PostgreSQL (mpp_sessions + mpp_payments tables from migration 075)
when DATABASE_URL is set. Falls back to in-memory for dev/demo.
"""
from __future__ import annotations

import logging
from datetime import UTC, datetime
from decimal import Decimal
from uuid import uuid4

from fastapi import APIRouter, Depends, HTTPException, Request, status

from server.authz import Principal, require_principal
from server.domains.mpp_session import (
    MPPBudgetExceededError,
    MPPSessionExpiredError,
    MPPSessionInactiveError,
    apply_payment_budget,
    ensure_card_budget,
    ensure_session_can_execute,
)
from server.models.mpp import (
    CreateMPPSessionRequest,
    ExecutePaymentRequest,
    ExecutePaymentResponse,
    IssueCardRequest,
    IssueCardResponse,
    MPPSessionResponse,
    PolicyEvaluateRequest,
    PolicyEvaluateResponse,
    mpp_session_response_from_record,
)
from server.repositories.mpp_session_repository import (
    close_session_record,
    create_session_record,
    deduct_memory_session,
    get_memory_session,
    load_session,
    mark_session_expired,
    record_payment,
)
from server.services.mpp_execution import execute_chain_payment
from server.services.mpp_virtual_cards import issue_mpp_virtual_card

logger = logging.getLogger(__name__)

router = APIRouter()


# ── Endpoints ──────────────────────────────────────────────────────────


@router.post("/sessions", response_model=MPPSessionResponse, status_code=status.HTTP_201_CREATED)
async def create_session(
    req: CreateMPPSessionRequest,
    principal: Principal = Depends(require_principal),
):
    """Create an MPP payment session with a spending limit."""
    from datetime import timedelta

    session_id = f"mpp_sess_{uuid4().hex[:16]}"
    now = datetime.now(UTC)
    expires_at_dt = None
    expires_at = None
    if req.expires_in_seconds:
        expires_at_dt = now + timedelta(seconds=req.expires_in_seconds)
        expires_at = expires_at_dt.isoformat()

    session = {
        "session_id": session_id,
        "org_id": principal.org_id,
        "mandate_id": req.mandate_id,
        "wallet_id": req.wallet_id,
        "agent_id": req.agent_id,
        "method": req.method,
        "chain": req.chain,
        "currency": req.currency,
        "spending_limit": req.spending_limit,
        "remaining": req.spending_limit,
        "total_spent": Decimal("0"),
        "payment_count": 0,
        "status": "active",
        "created_at": now,
        "closed_at": None,
        "expires_at": expires_at,
    }
    await create_session_record(
        session=session,
        metadata=req.metadata or {},
        expires_at_dt=expires_at_dt,
    )
    logger.info("MPP session created: %s (limit=%s %s)", session_id, req.spending_limit, req.currency)

    return mpp_session_response_from_record(
        session,
        next_steps=[
            "POST /api/v2/mpp/sessions/{session_id}/execute — Execute payment",
        ],
    )


async def _load_session(session_id: str, org_id: str) -> dict:
    """Load session from DB or memory, with org ownership check."""
    session = await load_session(session_id, org_id)
    if not session:
        raise HTTPException(status_code=404, detail="MPP session not found")
    if session["org_id"] != org_id:
        raise HTTPException(status_code=403, detail="Access denied")
    return session


@router.get("/sessions/{session_id}", response_model=MPPSessionResponse)
async def get_session(
    session_id: str,
    principal: Principal = Depends(require_principal),
):
    """Get MPP session status."""
    session = await _load_session(session_id, principal.org_id)
    return mpp_session_response_from_record(session)


@router.post("/sessions/{session_id}/execute", response_model=ExecutePaymentResponse)
async def execute_payment(
    session_id: str,
    req: ExecutePaymentRequest,
    request: Request,
    principal: Principal = Depends(require_principal),
):
    """Execute a payment within an MPP session.

    When a chain executor is available (SARDIS_CHAIN_MODE=live or simulated),
    this endpoint builds a PaymentMandate and executes a real Tempo transfer
    via the ChainExecutor. The resulting tx_hash is persisted to the
    mpp_payments record.
    """
    session = await _load_session(session_id, principal.org_id)

    try:
        ensure_session_can_execute(session)
        budget = apply_payment_budget(session, req.amount)
    except MPPSessionExpiredError as exc:
        await mark_session_expired(session_id)
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    except MPPSessionInactiveError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    except MPPBudgetExceededError as exc:
        raise HTTPException(
            status_code=422,
            detail=str(exc),
        ) from exc

    payment_id = f"mpp_pay_{uuid4().hex[:16]}"

    # ── Execute on-chain payment via ChainExecutor ────────────────────
    tx_hash: str | None = None
    pay_status = "completed"
    pay_error: str | None = None

    chain_executor = getattr(request.app.state, "chain_executor", None)
    if chain_executor:
        try:
            tx_hash = await execute_chain_payment(
                chain_executor=chain_executor,
                session=session,
                request=req,
                payment_id=payment_id,
                organization_id=principal.org_id,
            )

        except Exception as exc:
            pay_status = "failed"
            pay_error = str(exc)
            logger.error(
                "MPP on-chain payment failed: %s error=%s",
                payment_id, pay_error,
            )
            # Do NOT deduct budget on failure — raise so caller can retry
            raise HTTPException(
                status_code=502,
                detail=f"On-chain payment failed: {pay_error}",
            )
    else:
        logger.warning("MPP execute: no chain_executor on app.state — budget-only mode")

    await record_payment(
        session_id=session_id,
        amount=req.amount,
        merchant=req.merchant,
        merchant_url=req.merchant_url,
        metadata=req.metadata or {},
        payment_id=payment_id,
        currency=session["currency"],
        chain=session["chain"],
        new_remaining=budget.remaining,
        new_total=budget.total_spent,
        new_count=budget.payment_count,
        new_status=budget.status,
        pay_status=pay_status,
        tx_hash=tx_hash,
        created_at=datetime.now(UTC),
    )

    logger.info(
        "MPP payment executed: %s in session %s (amount=%s, remaining=%s, tx=%s)",
        payment_id, session_id, req.amount, budget.remaining, tx_hash,
    )

    return ExecutePaymentResponse(
        payment_id=payment_id,
        session_id=session_id,
        amount=str(req.amount),
        merchant=req.merchant,
        status=pay_status,
        tx_hash=tx_hash,
        chain=session["chain"],
        remaining=str(budget.remaining),
        next_steps=[
            "GET /api/v2/ledger/entries — Check audit trail",
            "POST /api/v2/mpp/sessions/{session_id}/close — Close session",
        ],
    )


@router.post("/sessions/{session_id}/close", response_model=MPPSessionResponse)
async def close_session(
    session_id: str,
    principal: Principal = Depends(require_principal),
):
    """Close an MPP session and settle remaining balance."""
    session = await _load_session(session_id, principal.org_id)

    if session["status"] not in ("active", "exhausted"):
        raise HTTPException(status_code=409, detail=f"Session is already {session['status']}")

    closed_at = datetime.now(UTC)

    await close_session_record(session_id, closed_at)

    session["status"] = "closed"
    session["closed_at"] = closed_at.isoformat()

    logger.info(
        "MPP session closed: %s (spent=%s, payments=%d)",
        session_id, session["total_spent"], session["payment_count"],
    )
    return mpp_session_response_from_record(session)


@router.post("/evaluate", response_model=PolicyEvaluateResponse)
async def evaluate_policy(
    req: PolicyEvaluateRequest,
    principal: Principal = Depends(require_principal),
):
    """Evaluate Sardis policy for an MPP payment (proxy to Sardis Guard)."""
    # Basic policy evaluation — in production, this calls the Sardis Guard service
    allowed = True
    reason = "ALLOWED by default policy"
    checks_passed = 12
    checks_total = 12

    # Basic limit checks
    if req.amount > Decimal("10000"):
        allowed = False
        reason = "Amount exceeds maximum single payment limit ($10,000)"
        checks_passed = 11

    logger.info(
        "MPP policy evaluation: amount=%s merchant=%s result=%s",
        req.amount, req.merchant, "ALLOWED" if allowed else "DENIED",
    )

    return PolicyEvaluateResponse(
        allowed=allowed,
        reason=reason,
        checks_passed=checks_passed,
        checks_total=checks_total,
    )


@router.post("/simulate", response_model=PolicyEvaluateResponse)
async def simulate_policy(
    req: PolicyEvaluateRequest,
    principal: Principal = Depends(require_principal),
):
    """Dry-run policy check without executing payment."""
    # Same as evaluate but explicitly labeled as simulation
    return await evaluate_policy(req, principal)


# ── Virtual Card Issuance via Laso/Locus MPP ─────────────────────────


@router.post("/cards/issue", response_model=IssueCardResponse, status_code=status.HTTP_201_CREATED)
async def issue_virtual_card(
    req: IssueCardRequest,
    principal: Principal = Depends(require_principal),
):
    """Issue a virtual prepaid card via Laso Finance MPP service.

    The card is a non-reloadable Visa prepaid card issued through
    the Locus MPP proxy. Amount must be between $5 and $1,000.

    If session_id is provided, the card amount will be deducted
    from the MPP session budget.

    Restrictions:
    - US-only (IP-locked)
    - Non-reloadable
    - No 3D Secure
    - Card amount must match checkout total exactly
    """
    # If session provided, check budget
    if req.session_id:
        session = get_memory_session(req.session_id)
        if not session:
            raise HTTPException(status_code=404, detail="MPP session not found")
        if session["org_id"] != principal.org_id:
            raise HTTPException(status_code=403, detail="Access denied")
        try:
            ensure_card_budget(session, req.amount)
        except MPPSessionInactiveError as exc:
            raise HTTPException(status_code=409, detail=str(exc)) from exc
        except MPPBudgetExceededError as exc:
            raise HTTPException(
                status_code=422,
                detail=str(exc),
            ) from exc

    try:
        response = await issue_mpp_virtual_card(req)
        if req.session_id:
            deduct_memory_session(req.session_id, req.amount)
        logger.info("MPP virtual card issued: %s amount=%s for org %s", response.card_id, req.amount, principal.org_id)
        return response
    except ImportError:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="sardis_mpp package not installed -- set SARDIS_CHAIN_MODE=simulated for sandbox cards",
        )
